### ISSUE TITLE
just record what i found

### ISSUE THREAD

**Issue by FelixAblazeSr** [Unlimited-Gold-Work/GANtest#1]
import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.gridspec as gridspec
import os
from tensorflow.examples.tutorials.mnist import input_data
from PIL import Image

sess = tf.InteractiveSession() #可在之後再設定參數 ft.session需要先定義完

mb_size = 3
Z_dim = 100

#mnist = input_data.read_data_sets('../../MNIST_data', one_hot=True)

photolist=[]
def ResizeAndConToNumpyArray(pic,X_size,Y_size):
    gray = pic.convert('L')
    gray_rz = gray.resize((X_size,Y_size), Image.ANTIALIAS)
    np_gray= np.array(gray_rz,float)
    np_gray = np_gray.ravel()
    np_gray /= 255
    return np_gray
def next_batch(num, data):
    '''
    Return a total of `num` random samples and labels. 
    '''
    idx = np.arange(0 , len(data))
    np.random.shuffle(idx)
    idx = idx[:num]
    data_shuffle = [data[ i] for i in idx]

    return np.asarray(data_shuffle)

for root,photo,files in os.walk("desert"):
    print(root)
    for f in files:
        print(os.path.join(root,f))
        im=Image.open(os.path.join(root,f))
        photolist.append(ResizeAndConToNumpyArray(im,28,28))

photolist=np.array(photolist)

def weight_var(shape, name):
    return tf.get_variable(name=name, shape=shape, initializer=tf.contrib.layers.xavier_initializer())
    #共享變數--變數名稱/變數維度/變數初始化方式(此處return為初始化權重矩阵)

def bias_var(shape, name):
    return tf.get_variable(name=name, shape=shape, initializer=tf.constant_initializer(0))
    #常數初始化函數

# discriminater net

X = tf.placeholder(tf.float32, shape=[None, 784], name='X') #類似變數暫存器--型態/維度/變數名

D_W1 = weight_var([784, 128], 'D_W1')
D_b1 = bias_var([128], 'D_b1')

D_W2 = weight_var([128, 1], 'D_W2')
D_b2 = bias_var([1], 'D_b2')

theta_D = [D_W1, D_W2, D_b1, D_b2]

# generator net

Z = tf.placeholder(tf.float32, shape=[None, 100], name='Z')

G_W1 = weight_var([100, 128], 'G_W1')
G_b1 = bias_var([128], 'G_B1')

G_W2 = weight_var([128, 784], 'G_W2')
G_b2 = bias_var([784], 'G_B2')

theta_G = [G_W1, G_W2, G_b1, G_b2]

def generator(z): #製造者
    G_h1 = tf.nn.relu(tf.matmul(z, G_W1) + G_b1) #矩陣a*矩陣b+矩陣c,relu小於0的變成0
    G_log_prob = tf.matmul(G_h1, G_W2) + G_b2 #同上
    G_prob = tf.nn.sigmoid(G_log_prob) #計算G_log_prob的sigmoid
    return G_prob

def discriminator(x): #鑑定者,函式同上
    D_h1 = tf.nn.relu(tf.matmul(x, D_W1) + D_b1)
    D_logit = tf.matmul(D_h1, D_W2) + D_b2
    D_prob = tf.nn.sigmoid(D_logit)
    return D_prob, D_logit

G_sample = generator(Z)
D_real, D_logit_real = discriminator(X) #訓練過程相反 鑑定者從多到少
D_fake, D_logit_fake = discriminator(G_sample) #製造者則是從少變多,最後要成為一張圖

#loss function
#tf.nn.sigmoid_cross_entropy_with_logits--M*X的矩陣/label(維度同logits) sigmoid建議二分法,另外如softmax則是二分以上(以機率和為一來做)
#tf.ones_like--全矩陣元素都設為一,反之為零
#tf.reduce_mean--取矩陣的平均值(平均值為一個數字,並非矩陣)
D_loss_real = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(
    logits=D_logit_real, labels=tf.ones_like(D_logit_real)))
D_loss_fake = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(
    logits=D_logit_fake, labels=tf.zeros_like(D_logit_fake)))
D_loss = D_loss_real + D_loss_fake
G_loss = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(
    logits=D_logit_fake, labels=tf.ones_like(D_logit_fake)))

#tf.train.AdamOptimizer--優化函數
#minimize()方法通過更新var_list来最小化loss損失函數
#minimize() = compute_gradients() + apply_gradients()>>>前者用來計算var_list中的變數,後者把前者得到的梯度傳播給weights和biases
D_optimizer = tf.train.AdamOptimizer().minimize(D_loss, var_list=theta_D)
G_optimizer = tf.train.AdamOptimizer().minimize(G_loss, var_list=theta_G)

def sample_Z(m, n):
    '''Uniform prior for G(Z)'''
    return np.random.uniform(-1., 1., size=[m, n]) #从一个均匀分布[low,high)中随机采样，注意定义域是左闭右开，即包含low，不包含high,size為輸出個數(m*n)

sess.run(tf.global_variables_initializer()) #初始化模型參數,雖然很重要,但是不了解為什麼 run激活後面的運算

for it in range(1000000):
    X_mb = next_batch(mb_size,photolist) #每批images,label  _為臨時變數,表示後面不會再用到,通常會在迴圈時使用
    _, D_loss_curr = sess.run([D_optimizer, D_loss],
                              feed_dict={X: X_mb, Z: sample_Z(mb_size, Z_dim)}) #run[a,b]執行a和b操作  feed_dict在這個函式範圍中把X用X_mb替換,以此類推
    _, G_loss_curr = sess.run([G_optimizer, G_loss],
                              feed_dict={Z: sample_Z(mb_size, Z_dim)}) #同上,不過G_loss不用用到兩個參數

    if it % 1000 == 0: #每1000次輸出一次製造者與鑑定者的loss率
        print('Iter: {}'.format(it))
        print('D loss: {:.4}'.format(D_loss_curr))
        print('G_loss: {:.4}'.format(G_loss_curr))
        print()

def sample_Z(m, n):
    '''Uniform prior for G(Z)'''
    return np.random.uniform(-1., 1., size=[m, n])
def plot(samples): #保存圖片
    fig = plt.figure(figsize=(4, 4))
    gs = gridspec.GridSpec(4, 4)
    gs.update(wspace=0.05, hspace=0.05)

    for i, sample in enumerate(samples):  # [i,samples[i]] imax=16
        ax = plt.subplot(gs[i])
        plt.axis('off')
        ax.set_xticklabels([])
        ax.set_aspect('equal')
        plt.imshow(sample.reshape(28, 28), cmap='Greys_r')

    return fig

if not os.path.exists('out2/'):
    os.makedirs('out2/')

#訓練製造者
sess.run(tf.global_variables_initializer())

i = 0
for it in range(1000000):
    if it % 1000 == 0: #每1000次保存16張圖片
        samples = sess.run(G_sample, feed_dict={
                           Z: sample_Z(16, Z_dim)})  # 16*784
        fig = plot(samples)
        plt.savefig('out2/{}.png'.format(str(i).zfill(3)), bbox_inches='tight')
        i += 1
        plt.close(fig)

    X_mb = next_batch(mb_size,photolist)

    _, D_loss_curr = sess.run([D_optimizer, D_loss], feed_dict={
                              X: X_mb, Z: sample_Z(mb_size, Z_dim)})
    _, G_loss_curr = sess.run([G_optimizer, G_loss], feed_dict={
                              Z: sample_Z(mb_size, Z_dim)})

    if it % 1000 == 0:
        print('Iter: {}'.format(it))
        print('D loss: {:.4}'.format(D_loss_curr))
        print('G_loss: {:.4}'.format(G_loss_curr))
        print()
